SAAS-67 Document go logging facility and prepare it for release (#6)
socifi/go-logging-facility: Readme.md

# Go logging facility

> Based on [Apex log](https://github.com/apex/log) with few tweaks. Unless otherwise noted (mainly in `pkg.go`), this package should be thread safe. Just for the record the architecture leaves almost all thread safety on handlers so if you are uncertain, consult the documentation and code of handlers you use.

Reviewer: ```suggestion
Based on [Apex log](https://github.com/apex/log) with few tweaks. Unless otherwise noted (mainly in `pkg.go`), this package should be thread safe. Just for the record the architecture leaves almost all thread safety on handlers so if you are uncertain, consult the documentation and code of the handlers you use.
```

## Functionality

### Autogenerated levels

> This package contains python script for log level automatic generation. The script contains default values used in SOCIFI but can be easily tweaked to generate entirely custom level names and level numbers. Just alter the level names and level codes at the top of the file `gen_levels.py` and run it. It will take care about everything else (unless you deleted guideline comments).

Reviewer: Would not use "The script contains default values used in SOCIFI" but rather "[RFC 5424](https://tools.ietf.org/html/rfc5424)". It's more generic and looks more professional :)
```suggestion
This package contains python script for log level automatic generation. By default, the script uses log levels that comply with RFC 5424 levels (debug, info, notice, warning, error, critical, alert, emergency) but can be easily tweaked to generate entirely custom level names and level numbers. Just alter the level names and level codes at the top of the file `gen_levels.py` and run it. It will take care about everything else (unless you deleted guideline comments).
```

### Hooks

> This package allows registering of hooks which will be run for every field in entry to check specific things in field name and then apply some function to it's content. It's incredibly useful e.g. for password hiding in logs. Example implementations of hooks are in folder `hook/` and amongst them is a dsn hook which will check if a field name contains `dsn` (case insensitive). This package will then try to convert dsn interface to string and hide password in it. Rationale behind this is that some informations might be very useful in troubleshooting some problems but logging passwords is very bad practice. This feature is entirely opt-in and might be included by importing module dsn as such:

Reviewer: ```suggestion
This package allows registering of hooks which will be run for every field in an entry to check specific things in field name and then apply some function to its content. It's incredibly useful e.g. for password hiding in logs. Example implementations of hooks are in folder `hook/` and amongst them is a dsn hook which will check if a field name contains `dsn` (case insensitive). This package will then try to convert dsn interface to string and hide password in it. The rationale behind this is that some information might be very useful in troubleshooting some problems but logging passwords is very bad practice. This feature is entirely opt-in and might be included by importing module dsn as such:
```

> This package allows registering of hooks which will be run for every field in entry to check specific things in field name and then apply some function to it's content. It's incredibly useful e.g. for password hiding in logs. Example implementations of hooks are in folder `hook/` and amongst them is a dsn hook which will check if a field name contains `dsn` (case insensitive). This package will then try to convert dsn interface to string and hide password in it. Rationale behind this is that some informations might be very useful in troubleshooting some problems but logging passwords is very bad practice. This feature is entirely opt-in and might be included by importing module dsn as such:

Reviewer: I would split it to make the message more clear:
1) It supports hooks (what is it, how to use it, ...)
2) DNS string sanitizer (what it is, why should one use it, how, ...)

Generally, try to avoid such long text. The text shall be as short and as easy to read as possible. Long texts are more likely to be skipped when reading. (at least use paragraphs :-) )

	// ...
	"github.com/socifi/go-logging-facility"
	_ "github.com/socifi/go-logging-facility/hook/dsn"
	// ...
)
```

#### How to write a hook

Hook is an interface with two functions. First one is `Check(string) bool` which takes field name as parameter and returns if the field should be processed or not. The other function is `Sanitize(interface{}) interface{}` which takes a field value and does some processing on it and returns processed interface which will be saved or sent to logging service.

```golang
package hook

import (
	"github.com/socifi/go-logging-facility"
	"strings"
)

type hook struct{}

var h hook

// Register hook to the mechanism
func init() {
	log.RegisterSanitizeHook(h)
}

func (dsn) Check(s string) bool {
	// string.Contains was chosen for simplicity, any logic can be here
	if strings.Contains(strings.ToLower(s), "password") {
		return true
	}
	return false
}

func (dsn) Sanitize(v interface{}) interface{} {
	// In case it's impossible to convert to chosen type return original value and do nothing with it
	s, ok := v.(string)
	if !ok {
		return v
	}

	// Again any logic (e.g. regex) can be here, this is for simplicity
	s = "***"

	return s
}
```

### Atexit

> Inspired by [atexit module](github.com/tebeka/atexit) go logging facility also offers possibility to register a function which will be run on system exit. However this functionality has the same caveat as atexit, the program needs to closed with any `Exit(code int)` function in this module. This function as well as `AddExitHandler(handler func())` which registers functions to be run on program exit is defined for both module, logger and entry.

Reviewer: ```suggestion
Inspired by [atexit module](github.com/tebeka/atexit) go logging facility also offers the possibility to register a function which will be run on system exit. However this functionality has the same caveat as atexit, the program needs to be closed with any `Exit(code int)` function in this module. This function as well as `AddExitHandler(handler func())` which registers functions to be run on program exit is defined for both module, logger and entry.
```

A side note: To access local variables use closures:

```golang
	// ...
	a := "something"
	log.AddExitHandler( func() {
		fmt.Println(a)
	})
	// ...

	// End the program
	log.Exit(0)
```

### Simple logger initialization from config json

> There is also initialization script which allows simple load of configuration needed by logger. This feature is now limited to json and elastic handlers and might be changed in the future to allow better interoperability with e.g. [Viper](https://github.com/spf13/viper). __We therefore cannot guarantee any backward compatibility in this module!__

Reviewer: ```suggestion
There is also initialization script which allows a simple load of configuration needed by the ogger. This feature is now limited to JSON and elastic handlers and might be changed in the future to allow better interoperability with e.g. [Viper](https://github.com/spf13/viper). __We, therefore, cannot guarantee any backward compatibility in this module!__
```

> But for now the example usage is following:

Reviewer: ```suggestion
But for now the example usage is the following:
```

```golang
package main

import (
	"encoding/json"
	"fmt"
	"github.com/socifi/go-logging-facility/handlers/init"
	_ "github.com/socifi/go-logging-facility/hook/dsn"
	"io/ioutil"
	"os"
)

type autoGenerated struct {
	/* Your other config here... */
	Logging loginit.LogConfig `json:"logging"`
}

func main() {
	raw, err := ioutil.ReadFile("./config.json")
	if err != nil {
		fmt.Println(err.Error())
		os.Exit(1)
	}

	var c autoGenerated
	json.Unmarshal(raw, &c)
	fmt.Printf("%#v\n", c)
	logger := loginit.Init(c.Logging)
	logger.WithField("dsn", "host=localhost port=5432 user=postgres password=postgres dbname=postgres sslmode=disable").Info("trial")
}
```

## Handlers

> Some handlers which support only fixed number of log levels have been discarded and only these are left. PR for simple text handler which is able to process any number of levels is welcome.

Reviewer: ```suggestion
Some handlers which support a fixed number of log levels only have been discarded and only the following were kept. PR for simple text handler which will be able to process any number of levels is welcome.
```
